File: aoc/2024/aoc202424.hs
module AOC202424 where

import Data.Bits(xor,(.&.),(.|.))
import Data.List(intercalate,sort)
import Data.Map(Map,alter,empty,findWithDefault,fromList,insert,member,(!))
import Data.Set(Set)
import qualified Data.Set

import AOC

aoc = AOC {
    day="24",
    aocTests=[
        AOCTest {
            testData=unlines [
                "x00: 1",
                "x01: 1",
                "x02: 1",
                "y00: 0",
                "y01: 1",
                "y02: 0",
                "",
                "x00 AND y00 -> z00",
                "x01 XOR y01 -> z01",
                "x02 OR y02 -> z02"
                ],
            testResult=Just "4",
            testResult2=Nothing
            },
        AOCTest {
            testData=unlines [
                "x00: 1",
                "x01: 0",
                "x02: 1",
                "x03: 1",
                "x04: 0",
                "y00: 1",
                "y01: 1",
                "y02: 1",
                "y03: 1",
                "y04: 1",
                "",
                "ntg XOR fgs -> mjb",
                "y02 OR x01 -> tnw",
                "kwq OR kpj -> z05",
                "x00 OR x03 -> fst",
                "tgd XOR rvg -> z01",
                "vdt OR tnw -> bfw",
                "bfw AND frj -> z10",
                "ffh OR nrd -> bqk",
                "y00 AND y03 -> djm",
                "y03 OR y00 -> psh",
                "bqk OR frj -> z08",
                "tnw OR fst -> frj",
                "gnj AND tgd -> z11",
                "bfw XOR mjb -> z00",
                "x03 OR x00 -> vdt",
                "gnj AND wpb -> z02",
                "x04 AND y00 -> kjc",
                "djm OR pbm -> qhw",
                "nrd AND vdt -> hwm",
                "kjc AND fst -> rvg",
                "y04 OR y02 -> fgs",
                "y01 AND x02 -> pbm",
                "ntg OR kjc -> kwq",
                "psh XOR fgs -> tgd",
                "qhw XOR tgd -> z09",
                "pbm OR djm -> kpj",
                "x03 XOR y03 -> ffh",
                "x00 XOR y04 -> ntg",
                "bfw OR bqk -> z06",
                "nrd XOR fgs -> wpb",
                "frj XOR qhw -> z04",
                "bqk OR frj -> z07",
                "y03 OR x01 -> nrd",
                "hwm AND bqk -> z03",
                "tgd XOR rvg -> z12",
                "tnw OR pbm -> gnj"
                ],
            testResult=Just "2024",
            testResult2=Nothing
            },
        AOCTest {
            testData=unlines [
                "x00: 0",
                "x01: 1",
                "x02: 0",
                "x03: 1",
                "x04: 0",
                "x05: 1",
                "y00: 0",
                "y01: 0",
                "y02: 1",
                "y03: 1",
                "y04: 0",
                "y05: 1",
                "",
                "x00 AND y00 -> z05",
                "x01 AND y01 -> z02",
                "x02 AND y02 -> z01",
                "x03 AND y03 -> z03",
                "x04 AND y04 -> z04",
                "x05 AND y05 -> z00"
                ],
            testResult=Nothing,
            testResult2=Nothing
            }
        ],
    aocCode=Code {
        codeParse=parse,
        codeParse2=parse,
        codeTest=result,
        codeTest2=undefined,
        codeResult=result,
        codeResult2=result2
        }
    }

data Op = IN Int
        | XOR String String String
        | OR  String String String
        | AND String String String
    deriving (Eq,Ord,Show)

parse :: String -> [(String,Op)]
parse = map p . filter (not . null) . map words . lines
  where
    p [a,b] = (init a,IN $ read b)
    p [a,"XOR",b,"->",c] = (c,XOR a b c)
    p [a,"OR", b,"->",c] = (c,OR  a b c)
    p [a,"AND",b,"->",c] = (c,AND a b c)

zs :: [(String,Op)] -> [String]
zs = sort . filter ((== 'z') . head) . map fst

result input = sum $ zipWith (*) (map (table!) $ zs input) [2^n | n <- [0..]]
  where
    table = fromList $ map (fmap op) input
    op (IN n) = n
    op (XOR a b _) = table!a `xor` table!b
    op (OR  a b _) = table!a .|.   table!b
    op (AND a b _) = table!a .&.   table!b

tabulate :: [(String,Op)]
         -> (Map String [Op],Map String [Op],Map String [Op],
             Map String Op,Set String)
tabulate = foldr collect (empty,empty,empty,empty,Data.Set.empty)
  where
    collect (_,op@(XOR a b c)) (xors,ors,ands,outs,ins) =
        (add op a b xors,ors,ands,insert c op outs,ins)
    collect (_,op@(OR  a b c)) (xors,ors,ands,outs,ins) =
        (xors,add op a b ors,ands,insert c op outs,ins)
    collect (_,op@(AND a b c)) (xors,ors,ands,outs,ins) =
        (xors,ors,add op a b ands,insert c op outs,ins)
    collect (v,_) (xors,ors,ands,outs,ins) =
        (xors,ors,ands,outs,Data.Set.insert v ins)
    add op a b =
        alter (Just . maybe [op] (op:)) a . alter (Just . maybe [op] (op:)) b

x00 = "x00"
y00 = "y00"
z00 = "z00"
next [c,tens,'9'] = [c,succ tens,'0']
next [c,tens,ones] = [c,tens,succ ones]

swapList :: (Map String [Op],Map String [Op],Map String [Op],
             Map String Op,Set String)
         -> [String]
swapList (xors,ors,ands,outs,ins) = checkAdd00
  where
    addSwap :: String -> String -> Map String String -> Map String String
    addSwap a b swaps = insert a b $ insert b a swaps
    swap :: String -> Map String String -> String
    swap a swaps = findWithDefault a a swaps

    checkZ :: String -> String -> Map String String -> Bool
    checkZ z out swaps = checkZout (outs!z)
      where
        checkZout (XOR _ _ outz) = out == swap outz swaps
        checkZout _ = False

    checkAdd00 :: [String]
    checkAdd00
      | xorX00 /= xorY00 = error (show (xorX00,xorY00))
      | checkZ z00 xorX00 empty = checkCarry00 empty
      | otherwise = z00 : xorX00 : checkCarry00 (addSwap z00 xorX00 empty)
      where
        (XOR _ _ xorX00:_) = xors!x00
        (XOR _ _ xorY00:_) = xors!y00

    checkCarry00 :: Map String String -> [String]
    checkCarry00 swaps
      | andX00 /= andY00 = error (show (andX00,andY00))
      | otherwise = checkAdd (next x00) (next y00) andX00 (next z00) swaps
      where
        (AND _ _ andX00:_) = ands!x00
        (AND _ _ andY00:_) = ands!y00

    checkAdd :: String -> String -> String -> String -> Map String String
             -> [String]
    checkAdd x y carry z swaps
      | not (Data.Set.member x ins) && z == carry = []
      | not (Data.Set.member x ins) = [carry,z]
        -- only (x XOR y) XOR carry in my input
        -- no cases of (x XOR carry) XOR y in my input
        -- no cases of (y XOR carry) XOR x in my input
      | swap xorX swaps /= swap xorY swaps = undefined
      | not (member (swap xorX swaps) xors)
              && xxory2 == swap carry swaps =
          carry2:xorX:checkAdd x y carry z (addSwap carry2 xorX swaps)
      | not (member (swap xorX swaps) xors)
              && carry2 == swap carry swaps =
          xxory2:xorX:checkAdd x y carry z (addSwap xxory2 xorX swaps)
      | not (member (swap carry swaps) xors)
              && xxory1 == swap carry swaps =
          carry1:carry:checkAdd x y carry z (addSwap carry1 carry swaps)
      | not (member (swap carry swaps) xors)
              && carry1 == swap carry swaps =
          xxory1:carry:checkAdd x y carry z (addSwap xxory1 carry swaps)
      | swap xorX swaps == swap xorY swaps
              && swap xorXorX swaps == swap xorCarry swaps
              && checkZ z xorXorX swaps =
          checkCarry x y carry z swaps
      | swap xorX swaps == swap xorY swaps
              && swap xorXorX swaps == swap xorCarry swaps =
          z:xorXorX:checkCarry x y carry z (addSwap z xorXorX swaps)
      | otherwise = undefined
      where
        (XOR _ _ xorX:_) = xors!x
        (XOR _ _ xorY:_) = xors!y
        (XOR xxory1 carry1 xorXorX:_) = xors!swap xorX swaps
        (XOR xxory2 carry2 xorCarry:_) = xors!swap carry swaps

    checkCarry :: String -> String -> String -> String -> Map String String
               -> [String]
    checkCarry x y carry z swaps
        -- only ((x XOR y) AND carry) OR (x AND y) in my input
        -- no cases of ((x XOR carry) AND y) OR (x AND carry) in my input
        -- no cases of ((y XOR carry) AND x) OR (y AND carry) in my input
      | swap xorX swaps /= swap xorY swaps = undefined
      | swap andX swaps /= swap andY swaps = undefined
      | swap andXorX swaps == swap andCarry swaps
              && swap andX swaps == swap andY swaps
              && swap orAndXorX swaps == swap orAndX swaps =
          checkAdd (next x) (next y) orAndX (next z) swaps
      | otherwise = undefined
      where
        (XOR _ _ xorX:_) = xors!x
        (XOR _ _ xorY:_) = xors!y
        (AND _ _ andXorX:_) = ands!swap xorX swaps
        (AND _ _ andCarry:_) = ands!swap carry swaps
        (AND _ _ andX:_) = ands!x
        (AND _ _ andY:_) = ands!y
        (OR _ _ orAndXorX:_) = ors!swap andXorX swaps
        (OR _ _ orAndX:_) = ors!swap andX swaps

result2 = sort . swapList . tabulate

File: aoc/2025/aoc202424.go
package main

import (
	"bytes"
	"fmt"
	"slices"
)

func init() {
	Register(&aoc202424{
		AOC: AOC{
			Day:           24,
			InputFilename: "../2024/input/24.txt",
			Tests: []Test{
				Test{
					Input: `x00: 1
x01: 1
x02: 1
y00: 0
y01: 1
y02: 0

x00 AND y00 -> z00
x01 XOR y01 -> z01
x02 OR y02 -> z02
`,
					Part1: "4",
					Part2: "",
				},
				Test{
					Input: `x00: 1
x01: 0
x02: 1
x03: 1
x04: 0
y00: 1
y01: 1
y02: 1
y03: 1
y04: 1

ntg XOR fgs -> mjb
y02 OR x01 -> tnw
kwq OR kpj -> z05
x00 OR x03 -> fst
tgd XOR rvg -> z01
vdt OR tnw -> bfw
bfw AND frj -> z10
ffh OR nrd -> bqk
y00 AND y03 -> djm
y03 OR y00 -> psh
bqk OR frj -> z08
tnw OR fst -> frj
gnj AND tgd -> z11
bfw XOR mjb -> z00
x03 OR x00 -> vdt
gnj AND wpb -> z02
x04 AND y00 -> kjc
djm OR pbm -> qhw
nrd AND vdt -> hwm
kjc AND fst -> rvg
y04 OR y02 -> fgs
y01 AND x02 -> pbm
ntg OR kjc -> kwq
psh XOR fgs -> tgd
qhw XOR tgd -> z09
pbm OR djm -> kpj
x03 XOR y03 -> ffh
x00 XOR y04 -> ntg
bfw OR bqk -> z06
nrd XOR fgs -> wpb
frj XOR qhw -> z04
bqk OR frj -> z07
y03 OR x01 -> nrd
hwm AND bqk -> z03
tgd XOR rvg -> z12
tnw OR pbm -> gnj
`,
					Part1: "2024",
					Part2: "",
				},
			},
		},
	})
}

type aoc202424 struct {
	AOC
}

func (aoc *aoc202424) parse(input *Input) (map[string]bool, map[string][3]string) {
	wires := map[string]bool{}
	paragraph, _ := input.Paragraph()
	in := InputString(paragraph)
	for {
		w, _ := in.Word()
		n, ok := in.Word()
		if !ok {
			break
		}
		wires[w[:len(w)-1]] = n == "1"
	}
	gates := map[string][3]string{}
	for {
		a, _ := input.Word()
		op, _ := input.Word()
		b, _ := input.Word()
		input.Word()
		wire, ok := input.Word()
		if !ok {
			break
		}
		gates[wire] = [3]string{min(a, b), op, max(a, b)}
	}
	return wires, gates
}

func (aoc *aoc202424) wire(wires map[string]bool, gates map[string][3]string, wire string) bool {
	if v, ok := wires[wire]; ok {
		return v
	}
	gate := gates[wire]
	a := aoc.wire(wires, gates, gate[0])
	b := aoc.wire(wires, gates, gate[2])
	v := false
	switch gate[1] {
	case "AND":
		v = a && b
	case "OR":
		v = a || b
	case "XOR":
		v = a != b
	}
	wires[wire] = v
	return v
}

func (aoc *aoc202424) Part1(input *Input) string {
	wires, gates := aoc.parse(input)
	result := 0
	bit := 1
	for z := range len(gates) {
		wire := fmt.Sprintf("z%02d", z)
		if _, ok := gates[wire]; !ok {
			break
		}
		if aoc.wire(wires, gates, wire) {
			result |= bit
		}
		bit <<= 1
	}
	return IntResult(result)
}

func (aoc *aoc202424) Part2(input *Input) string {
	wires, gates := aoc.parse(input)
	nbits := len(wires) / 2

	// in my input, x and y only appear as X AND Y and X XOR Y
	ands := make([]string, nbits)
	xors := make([]string, nbits)
	andIndex := map[string]int{}
	xorIndex := map[string]int{}
	for k, v := range gates {
		if v[0][0] == 'x' && v[2][0] == 'y' {
			bit, _ := InputString(v[0][1:]).Int()
			if v[1] == "AND" {
				ands[bit] = k
				andIndex[k] = bit
			} else if v[1] == "XOR" {
				xors[bit] = k
				xorIndex[k] = bit
			}
		}
	}

	swaps := map[string]bool{}
	if "z00" != xors[0] {
		swaps["z00"] = true
		swaps[xors[0]] = true
	}

	// z = carryIn xor (x xor y)
	// carryOut = (x and y) or (carryIn and (x xor y))

	carryIns := make([]string, nbits)
	for k, v := range gates {
		if v[1] != "XOR" {
			continue
		}
		if _, ok := xorIndex[k]; ok {
			continue
		}
		carryIn := v[2]
		i, ok := xorIndex[v[0]]
		if !ok {
			carryIn = v[0]
			i, ok = xorIndex[v[2]]
			if !ok {
				continue
			}
		}
		carryIns[i] = carryIn
		z := fmt.Sprintf("z%02d", i)
		if z != k {
			swaps[z] = true
			swaps[k] = true
		}
	}
	if carryIns[1] != ands[0] {
		swaps[carryIns[1]] = true
		swaps[ands[0]] = true
	}

	carryAnds := make([]string, nbits)
	for k, v := range gates {
		if v[1] != "AND" {
			continue
		}
		if _, ok := andIndex[k]; ok {
			continue
		}
		carryIn := v[2]
		i, ok := xorIndex[v[0]]
		if !ok {
			carryIn = v[0]
			i, ok = xorIndex[v[2]]
			if !ok {
				continue
			}
		}
		carryAnds[i] = k
		if carryIns[i] == "" {
			panic("bad data")
		} else if carryIns[i] != carryIn {
			swaps[carryIns[i]] = true
			swaps[carryIn] = true
		}
	}

	carryOuts := make([]string, nbits)
	carryOuts[0] = ands[0]
	for k, v := range gates {
		if v[1] != "OR" {
			continue
		}
		carryAnd := v[2]
		i, ok := andIndex[v[0]]
		if !ok {
			carryAnd = v[0]
			i, ok = andIndex[v[2]]
			if !ok {
				continue
			}
		}
		carryOuts[i] = k
		if carryAnds[i] == "" {
			panic("bad data")
		} else if carryAnds[i] != carryAnd {
			swaps[carryAnds[i]] = true
			swaps[carryAnd] = true
		}
	}
	zlast := fmt.Sprintf("z%02d", nbits)
	if carryOuts[nbits-1] != zlast {
		swaps[zlast] = true
		swaps[carryOuts[nbits-1]] = true
	}

	for i := range nbits {
		if i == 0 || carryIns[i] != "" || carryOuts[i-1] == "" {
			continue
		}
		carryIn := carryOuts[i-1]
		for _, v := range gates {
			j := -1
			if v[0] == carryIn {
				j = 2
			} else if v[2] == carryIn {
				j = 0
			}
			if j < 0 {
				continue
			}
			swaps[xors[i]] = true
			swaps[v[j]] = true
			break
		}
	}

	s := []string{}
	for k := range swaps {
		s = append(s, k)
	}
	slices.Sort(s)
	buf := &bytes.Buffer{}
	for _, swap := range s {
		buf.WriteByte(',')
		buf.WriteString(swap)
	}
	buf.ReadByte()
	return buf.String()
}
